Scripts/Hue Wheel AIDA64.lua

```diff
@@ -0,0 +1,147 @@
+-- Copyright © 2018, Pedro Fanha
+--
+-- Permission to use, copy, modify, and/or distribute this software for any
+-- purpose with or without fee is hereby granted, provided that the above1
+-- copyright notice and this permission notice appear in all copies.
+--
+-- THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
+-- REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
+-- AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
+-- INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
+-- LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE
+-- OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
+-- PERFORMANCE OF THIS SOFTWARE.
+
+-- cpu, gpu, fan or hue
+mode = 'hue'
+
+-- Min and Max speed of the CPU fan
+fan = {750, 1800}
+-- temperatur minimum and max, applies for gpu and cpu
+temp = {35, 65}
+
+color_min = 0
+color_max = 200
+
+cpu_temp = 50
+gpu_temp = 60
+mb_temp = 50
+
+cpu_r = 0
+cpu_g = 0
+cpu_b = 255
+
+gpu_r = 0
+gpu_g = 255
+gpu_b = 0
+
+mb_r = 255
+mb_g = 255
+mb_b = 255
+
+color_steps = 0.5
+delay = 500
+
+
+-- https://gist.github.com/GigsD4X/8513963
+function HSVToRGB(hue, saturation, value)
+    -- Returns the RGB equivalent of the given HSV-defined color
+    -- (adapted from some code found around the web)
+
+    -- Get the hue sector
+    local hue_sector = math.floor( hue / 60 );
+    local hue_sector_offset = ( hue / 60 ) - hue_sector;
+
+    local p = value * ( 1 - saturation );
+    local q = value * ( 1 - saturation * hue_sector_offset );
+    local t = value * ( 1 - saturation * ( 1 - hue_sector_offset ) );
+
+    if hue_sector == 0 then
+        return value, t, p;
+    elseif hue_sector == 1 then
+        return q, value, p;
+    elseif hue_sector == 2 then
+        return p, value, t;
+    elseif hue_sector == 3 then
+        return p, q, value;
+    elseif hue_sector == 4 then
+        return t, p, value;
+    elseif hue_sector == 5 then
+        return value, p, q;
+    end;
+end;
+
+-- Adapted from nagisa/msi-rgb's Hue Wheel effect
+Lighting.SetStepDuration(511)
+Lighting.SetFlashingSpeed(0)
+Lighting.SetBreathingModeEnabled(false)
+
+function Color(value, temp1, temp2)
+    percent = (value - temp1) / (temp2 - temp1)
+    color = color_max - percent * (color_max - color_min)
+    return color
+end
+
+local color = 0
+while true do
+    if Aida64.IsInstalledAndRunning() then
+        if mode == 'fan' then
+            rpm = Aida64.GetSensorValue('FCPU')
+            color = Color(rpm, fan[1], fan[2])
+        elseif mode == 'cpu' then
+            hw_temp = Aida64.GetSensorValue('TCPU')
+            color = Color(hw_temp, temp[1], temp[2])
+        elseif mode == 'gpu' then
+            hw_temp = Aida64.GetSensorValue('TGPU1DIO')
+            color = Color(hw_temp, temp[1], temp[2])
+        else
+            while Aida64.GetSensorValue('TCPU') >= cpu_temp  do
```

Comment:
> This while loop seems redundant (and also forgets about Aida64.IsInstalledAndRunning()). A simple if should suffice, considering you're already within the main loop.

With an if, the normal "hue-color" would be set, but it's not supposed to do that.

> Is this intended?

Doing two or three effects of that at the same time would be derpy and it would be hard to identify which hardware is too hot.

Reply:
> With an if, the normal "hue-color" would be set, but it's not supposed to do that.
Yes, but the way you're doing it is kinda like a hack. You can take my other suggestion and move the non-hue modes into their separate function. Cleaner code and that problem will be solved as well.

> I can't imagine how I am supposed to check if AIDA64 is running while a while is depending on it... Ideas?
Well, in that case it would look something like:

```
else
    if curr_cpu_temp >= cpu_temp then
        -- ...
    else if ... then
        -- ...
    ...
    end
end
...
``Basically, the way you're doing it, the script will crash if it's stuck in the inner while loop and AIDA64 is stopped because the call to IsInstalledAndRunning is only done in the outer loop.

But if you move it into a separate function, none of that will be needed.

> Doing two or three effects of that at the same time would be derpy and it would be hard to identify which hardware is too hot.
I am aware, I just wondering if that was intended.

Reply:
> Basically, the way you're doing it, the script will crash if it's stuck in the inner while loop and AIDA64 is stopped in the meantime because the call to IsInstalledAndRunning is only done in the outer loop.

Yes, I know. I will try to fix this and make seperate functions. Thanks!